### PR TITLE
Flight tutorials

### PR DESCRIPTION
**Pull request by benarmston**
@mjtko / @jamesremuscat I'm not sure who best to direct this to.  Would one of you please take a look and let me know if this is heading in the right direction?

Also, I'd appreciate being pointed to instructions on how to test this.

Note: The `sysv` init file is unlikely to work.  I need to figure out how to deal with `PID_FILE` or lack of one.

Thanks.

### CODE REVIEW

File: alces-flight-tutorials/etc/alces-flight-www/tutorial.conf.tpl [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    access_log /var/log/alces-flight-www/tutorial-websockets-access.log;
+    error_log  /var/log/alces-flight-www/tutorial-websockets-error.log;
+}
+
+location /tutorial {
+    try_files $uri @tutorial;
+}
+
+location @tutorial {
+    proxy_pass http://127.0.0.1:3000;
```
Comment (mjtko) [inline alces-flight-tutorials/etc/alces-flight-www/tutorial.conf.tpl]: I think we should avoid port 3000, as it's a "common" port.  Perhaps throw it up in the 25000s somewhere? Say, 25288?

File: alces-flight-tutorials/init/sysv/clusterware-alces-flight-tutorials.el6 [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+LOG_FILE=/var/log/flight-tutorials/server.log
+DAEMON_HOME=/opt/tutorial
+DAEMON_CMD="su ${admin_user} -c 'yarn run start' ${LOG_FILE} 2>&1 &"
+
+start()
+{
+  echo -n Starting Alces Flight Tutorials Server:
+  cd $DAEMON_HOME
+  $DAEMON_CMD
+  sleep 2
```
Comment (mjtko) [inline alces-flight-tutorials/init/sysv/clusterware-alces-flight-tutorials.el6]: I would imagine that inserting this the `sleep` would be sufficient to generate the PID file:

```
echo "$!" > "${PID_FILE}"
```

Don't worry too much about the sysvinit script, but it's good to have a reasonable effort for completeness!

File: alces-flight-tutorials/metadata.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
+---
+install:
+  el6: |
+    yum install -y -e0 nodejs npm yarn
+  el7: |
+    yum install -y -e0 nodejs npm yarn
```
Comment (mjtko) [inline alces-flight-tutorials/metadata.yml]: I suspect none of these packages are available in EL7 and you'll need to enable EPEL.  e.g. https://github.com/alces-software/clusterware-sessions/blob/master/xfce/metadata.yml#L14-L21

Looks like there's no `yarn` package in EPEL.  I guess that's why you need `npm`, so `yarn` can be installed?  Is `yarn` required after build or can we use `npm run` or something instead?

File: alces-flight-tutorials/metadata.yml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  ubuntu1604: |
+    apt-get install -y -e0 nodejs npm yarn
+  _: |
+    require serviceware
+    if serviceware_is_installed alces-flight-www; then
+        cp etc/alces-flight-www/tutorial.conf.tpl \
+          "${cw_ROOT}"/etc/alces-flight-www/server-https.d/tutorial.conf
+        # XXX Does this work with relative paths?
```
Comment (mjtko) [inline alces-flight-tutorials/metadata.yml]: I don't remember, but I suspect not.  You might get some mileage out of:

`$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)` to get the absolute path.

File: alces-flight-tutorials/metadata.yml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            init/systemd/clusterware-alces-flight-tutorials.service \
+            > /etc/systemd/system/clusterware-alces-flight-tutorials.service
+        systemctl enable clusterware-alces-flight-tutorials.service
+    }
+  _: |
+    enable_alces_flight_tutorials
+
+build:
+  el6: |
+    yum install -y -e0 nodejs yarn
```
Comment (mjtko) [inline alces-flight-tutorials/metadata.yml]: EPEL required here too I reckon.

File: alces-flight-tutorials/metadata.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    tar -C /opt/tutorials --strip-components=1 -xzf /tmp/flight-tutorials-server.tar.gz
+
+    curl -L \
+      "https://github.com/alces-software/flight-tutorials-client/tarball/master" \
+      -o /tmp/flight-tutorials-client.tar.gz
+    mkdir -p /tmp/flight-tutorials-client
+    tar -C /tmp/flight-tutorials-client --strip-components=1 -xzf /tmp/flight-tutorials-client.tar.gz
+
+    pushd /tmp/flight-tutorials-client
+    yarn run build
```
Comment (mjtko) [inline alces-flight-tutorials/metadata.yml]: Think you might need to install `yarn` via `npm` first.

File: alces-flight-tutorials/resource/alces-flight-www/tutorial/config.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+---
+title: Flight Compute Tutorial
+short_title: Tutorial
+paras:
+ - >-
+   Get up and running with this simple Pegamoose-infused Alces Flight
```
Comment (mjtko) [inline alces-flight-tutorials/resource/alces-flight-www/tutorial/config.yml]: This copy could do with being updated. 😄

**Review by mjtko — COMMENTED** [review]
On the right track. Few comments inline.

For testing, you can spin up a Vagrant VM, or go the whole hog and build an AMI.

Vagrant VM: @bobwhitelock wrote some notes here: https://github.com/alces-software/imageware/blob/master/support/vagrant/DEVELOPMENT.md
AMI: you'll need to create a branch of `clusterware` which changes `scripts/registry.rc` to point at the branch of `clusterware-services` you want to test. Then build as usual using the `ami-creator` tool here: https://github.com/alces-software/imageware/tree/develop/support/aws

I'd start with the VM approach and graduate to the AMI approach once you're more confident that the build is correct.

Note that, until you've built and uploaded a tarball to the S3 bucket that contains the clusterware service tarballs, you'll need to use `alces service build` rather than `alces service install`.

**Comment by benarmston** [comment]
@mjtko a gazillion commits later and I think that this is done and ready for review/merge.

**Pull request merged**